Document javascript: URLs (#35599)
mdn/content: files/en-us/web/uri/schemes/javascript/index.md

---
title: "javascript: URLs"
slug: Web/URI/Schemes/javascript
page-type: guide
spec-urls: https://html.spec.whatwg.org/multipage/browsing-the-web.html#the-javascript:-url-special-case
---

{{QuickLinksWithSubpages("/en-US/docs/Web/URI")}}

> [!WARNING]

> > Using `javascript:` URLs on the web is discouraged as it may lead to execution of arbitrary code, similar to the ramifications of using [`eval()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval). It may also be inaccessible because it deviates from normal link behavior.

Reviewer: What do you mean by inaccessible here?

Author: I means "bad in terms of accessibility". Is there a better phrase?

Reviewer: Almost exactly that. But easiest to link to a definition.

```suggestion
> Using `javascript:` URLs on the web is discouraged as it may lead to execution of arbitrary code, similar to the ramifications of using [`eval()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval). It may also reduce {{glossary("accessibility")}} because it deviates from normal link behavior.
```

> `javascript:` URLs, URLs prefixed with the `javascript:` scheme, are used as if they were a kind of navigation target but actually executes JavaScript when the browser goes to the target location.

Reviewer: Follow the model of Data urls

```suggestion
**JavaScript URLs**, URLs prefixed with the `javascript:` scheme, are used as if they were a kind of navigation target, but actually execute JavaScript when the browser goes to the target location.
```

Reviewer: The description is not wrong ^^^ , but it also feels a little clunky, because I think it only navigates if the javascript evaluates to something that can be navigated to. It is perhaps sufficient for an intro, but perhaps we can tidy?

```suggestion
**JavaScript URLs**, URLs prefixed with the `javascript:` scheme, are used in navigation targets, but execute JavaScript before the browser attempts to navigate.
```

Author: I started with this version, but "JavaScript URL" is sort of ambiguous, and the spec also uses "`javascript:` URL". I'm more inclined to rename "Data URLs" to "data: URLs" but I'm on the fence for that because "Data URL" is less ambiguous.

Author: "JavaScript URL" is a fake navigation target. It is part of the navigation algorithm but it's possible that no "navigation" happens at all, unless the script evaluates to a string.

Reviewer: The rendering is ugly and hard to parse:

![image](https://github.com/user-attachments/assets/a1ee0fe4-7823-4968-99a7-f9805568a6f1)

It isn't ambiguous because you're providing the context in this page. If you're "randomly" referring to JavaScript URLs then sure - elsewhere we should refer to `javascript:` URLs.

Reviewer: But that's my opinion.

Reviewer: Maybe

```suggestion
**JavaScript URLs**, URLs prefixed with the `javascript:` scheme, are used as fake navigation targets that execute JavaScript when the browser attempts to navigate.
If the URL evaluates to a string, it is treated as HTML and rendered by the browser.
```

## Syntax

> `javascript:` URLs start with the `javascript:` scheme and are followed by JavaScript code. The code will be parsed as a script.

Reviewer: ```suggestion
**JavaScript URLs** start with the `javascript:` scheme and are followed by JavaScript code. The code will be parsed as a script.
```

```url
javascript:<script>
```

## Description

> `javascript:` URLs can be used anywhere a URL is a navigation target. This includes, but is not limited to:

Reviewer: OK to use the code form here, because we're not repeating ourselves after.

- The [`href`](/en-US/docs/Web/HTML/Element/a#href) attribute of an `<a>` or `<area>` element.
- The [`action`](/en-US/docs/Web/HTML/Element/form#action) attribute of a `<form>` element.
- The [`src`](/en-US/docs/Web/HTML/Element/iframe#src) attribute of an `<iframe>` element.
- The [`window.location`](/en-US/docs/Web/API/Window/location) JavaScript property.

> [!NOTE]
> Some other contexts that use URLs, such as the [`href`](/en-US/docs/Web/HTML/Element/link#href) attribute of `<link>` elements, do not allow `javascript:` URLs, because they are resource locations, not navigation targets. For these cases, if you want to write JavaScript inline, use [`data:`](/en-US/docs/Web/URI/Schemes/data) URLs with the `text/javascript` MIME type.

> When a browser attempts to navigate to such a location, it parses and executes the script body. The script may have a _completion value_ (not a return value), which is the same value if the script were executed with [`eval()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval). If the last statement is an [expression](/en-US/docs/Web/JavaScript/Reference/Statements/Expression_statement), the completion value is the value of that expression. If this completion value is a string, that string is treated as an HTML document and the browser navigates to a new page with that content, using the same URL as the current page. If the completion value is not a string, the browser only executes the code and does not navigate. Therefore, it's often recommended that if the script ends with a function call like `javascript:foo()`, you should prefix it with {{jsxref("Operators/void", "void")}} to prevent accidental navigation if the function happens to return a string.

Reviewer: not wrong, but I'd separate into two paragraphs for readability.

```suggestion
When a browser attempts to navigate to such a location, it parses and executes the script body. The script may have a _completion value_ (not a return value), which is the same value if the script were executed with [`eval()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval).

If the last statement is an [expression](/en-US/docs/Web/JavaScript/Reference/Statements/Expression_statement), the completion value is the value of that expression. If this completion value is a string, that string is treated as an HTML document and the browser navigates to a new page with that content, using the same URL as the current page. If the completion value is not a string, the browser only executes the code and does not navigate. Therefore, it's often recommended that if the script ends with a function call like `javascript:foo()`, you should prefix it with {{jsxref("Operators/void", "void")}} to prevent accidental navigation if the function happens to return a string.
```

Author: FYI, I've reverted this change. The paragraph flows like: "the browser runs the code, the code may have a completion value, the completion value is determined by blah blah, the completion value can do blah blah". So it's hard to find a natural place to break the paragraph.

Reviewer: True. I thought this was "natural enough". Not a hill I'd die on though.

```html example-good
<button id="btn">Click me</button>
<script>
  document.getElementById("btn").addEventListener("click", () => {
    alert("Hello, world!");
  });
</script>
```

In this example, the `href` attribute of an `<a>` element is set to a `javascript:` URL that navigates to a new page with the content "Hello, world!":

```html example-bad
<a href="javascript:pageContent">Click me</a>
<script>
  // Use a var so it becomes a global variable and can be read elsewhere
  var pageContent = "Hello, world!";
</script>
```

Note that because `javascript:` URLs do not create history entries, there's no way to go back to the previous page without refreshing.

### Using `javascript:` URLs as form actions

In this example, the `action` attribute of a `<form>` element is set to a `javascript:` URL that alerts a message when submitted:

```html example-bad
<form action="javascript:alert(myInput.value)">
  <input id="myInput" />
  <input type="submit" value="Submit" />
</form>
```

> Instead of doing this, consider listening for the form's `submit` event and handling it with JavaScript:

Reviewer: For all cases of ""Instead of doing this, consider" - Why?

Perhaps word as "Generally it is better to Xxxx, because YYYY". In some cases it it obvious, but even so worth spelling out.

Author: We are showing all ways you may use `javascript:` and suggesting alternatives, because as said in the intro there's probably no reason you should use `javascript:`, in the same way that you should never use inline HTML event handlers.

Reviewer: Fair enough. I guess I just prefer the previous cases where you were explicit - such as "This is a bad practice because the link is actually not a hyperlink. Consider making it a button instead:"
